ftw/publisher/core/tests/test_constrain_types_adapter.py

```diff
@@ -0,0 +1,43 @@
+from ftw.builder import Builder
+from ftw.builder import create
+from ftw.publisher.core.interfaces import IDataCollector
+from ftw.publisher.core.testing import PUBLISHER_CORE_INTEGRATION_TESTING
+from Products.CMFPlone.interfaces.constrains import ENABLED
+from Products.CMFPlone.interfaces.constrains import IConstrainTypes
+from Products.CMFPlone.interfaces.constrains import ISelectableConstrainTypes
+from unittest import skipUnless
+from unittest import TestCase
+from ftw.testing import IS_PLONE_5
+from zope.component import getAdapter
+
+
+@skipUnless(IS_PLONE_5, 'Test constrain types adapter for Plone 5')
+class TestConstrainTypesAdapter(TestCase):
+
+    layer = PUBLISHER_CORE_INTEGRATION_TESTING
+
+    def test_data_getter(self):
+        folder = create(Builder('folder').titled(u'The Folder'))
+        constrain_types = ISelectableConstrainTypes(folder)
+        constrain_types.setConstrainTypesMode(ENABLED)
+        types = ['Collection', 'Folder']
+        constrain_types.setLocallyAllowedTypes(types)
+        constrain_types.setImmediatelyAddableTypes(types)
+
+        adapter = getAdapter(folder, IDataCollector, name='constrain_types_adapter')
+
+        expected_data = {'mode': ENABLED, 'locally_allowed': types, 'immediately_addable': types}
+        self.assertEquals(expected_data, adapter.getData())
+
+    def test_data_setter(self):
+        folder = create(Builder('folder').titled(u'The Folder'))
+
+        adapter = getAdapter(folder, IDataCollector, name='constrain_types_adapter')
+        types = ['Document', 'Folder']
+        adapter.setData({'mode': ENABLED, 'locally_allowed': types, 'immediately_addable': types},
+                        metadata=None)
+
+        constrain_types = IConstrainTypes(folder)
+        self.assertEquals(ENABLED, constrain_types.getConstrainTypesMode())
+        self.assertEquals(types, constrain_types.getLocallyAllowedTypes())
+        self.assertEquals(types, constrain_types.getImmediatelyAddableTypes())
```

Comment:
Great! Perhaps it would make sense to put different Data to `constrain_types.getLocallyAllowedTypes()` and `constrain_types.getImmediatelyAddableTypes()` to validate that these two informations are different.

Reply:
Good idea 👍🏻